packages/react-components/react-utilities/src/compose/isSlot.ts

```diff
@@ -0,0 +1,62 @@
+import { isValidElement } from 'react';
+import type { ExtractSlotProps, Slot, SlotShorthandValue, UnknownSlotProps } from './types';
+
+/**
+ * Guard method that validates if a shorthand is a slot
+ * can be used to extends properties provided by a slot
+ *
+ * @example
+ * ```
+ * const backdropSlot = resolveShorthand(backdrop, {
+ *  defaultProps: {
+ *    onClick: useEventCallback(event => {
+ *     if (isSlot(backdrop)) {
+ *        backdrop.onClick?.(event)
+ *      }
+ *      // do something after passing click down the line
+ *    }),
+ *  },
+ * })
+ * ```
+ * @example
+ * ```
+ * const handleBackDropClick = (event) => {
+ *  // do your thing
+ * }
+ * const backdropSlot = resolveShorthand(backdrop, {
+ *  defaultProps: {
+ *    onClick: useEventCallback(mergeCallbacks(isSlot(backdrop) ? backdrop.onClick : undefined, handleBackdropClick))
+ * })
+ * ```
+ */
+export function isSlot<Shorthand extends Slot<UnknownSlotProps>>(
+  shorthand?: Shorthand,
+): shorthand is ExtractSlotProps<Shorthand> {
+  return (
+    shorthand !== null &&
+    typeof shorthand !== 'string' &&
+    typeof shorthand !== 'number' &&
+    !Array.isArray(shorthand) &&
+    !isValidElement(shorthand) &&
+    typeof shorthand === 'object'
```

Comment:
Checks for `string` & `number` are redundant there as we have a check for `object` later. I suggest to regroup conditions so it will be more obvious:

```suggestion
    typeof shorthand === 'object' &&
    shorthand !== null &&
    !Array.isArray(shorthand) &&
    !isValidElement(shorthand)
```